js/packages/openinference-instrumentation-langchain/src/tracer.ts

```diff
@@ -0,0 +1,107 @@
+import { BaseTracer, Run } from "@langchain/core/tracers/base";
+import {
+  Tracer,
+  SpanKind,
+  Span,
+  context,
+  trace,
+  SpanStatusCode,
+} from "@opentelemetry/api";
+import { isTracingSuppressed } from "@opentelemetry/core";
+import { SemanticConventions } from "@arizeai/openinference-semantic-conventions";
+import {
+  safelyFlattenAttributes,
+  safelyFormatIO,
+  safelyFormatInputMessages,
+  safelyFormatOutputMessages,
+  safelyGetOpenInferenceSpanKindFromRunType,
+} from "./utils";
+
+type RunWithSpan = {
+  run: Run;
+  span: Span;
+};
+
+export class LangChainTracer extends BaseTracer {
+  private tracer: Tracer;
+  private runs: Record<string, RunWithSpan | undefined> = {};
+  constructor(tracer: Tracer) {
+    super();
+    this.tracer = tracer;
+  }
+  name: string = "OpenInferenceLangChainTracer";
+  protected persistRun(_run: Run): Promise<void> {
+    return Promise.resolve();
+  }
+
+  protected async _startTrace(run: Run) {
+    await super._startTrace(run);
+    if (isTracingSuppressed(context.active())) {
+      return;
+    }
+
+    let activeContext = context.active();
+    const parentCtx = this.getParentSpanContext(run);
+    if (parentCtx != null) {
+      activeContext = trace.setSpanContext(context.active(), parentCtx);
+    }
```

Comment:
this bit is a tad confusing to me and requires a bit of explanation I think.

Reply:
will add some comments